### ISSUE TITLE
Make packages know about their source target

### ISSUE THREAD

**Issue by charlesseizilles** [xmake-io/xmake#4113]
### Is your feature request related to a problem? Please describe.

We have several softwares that depends on a common static library whose package is declared in our own repository.
This library has a lot of dependencies that evolve frequently and we don't have an easy way to expose them and their precise versions to the softwares.

Right now, each software have `add_requires`s for the dependency's dependencies in their `xmake.lua`.
An alternative is to declare dependencies in the package using `add_deps`, but then we would have to add a lot of branches to set the correct dependencies versions for each version of the library which would be worse for maintainability.
Another alternative we considered is not relying on xmake and copy the dependencies on installation of the library, but then they will be duplicated and we also rely on the matlab runtime which cannot be installed as a package and should not be copied on installation.

### Describe the solution you'd like

Since all dependencies are known in the library's `xmake.lua`, it would be simpler if the package could extract all the infromations it needs from it.
It could be automatic, for instance if there is a `xmake.lua` in the cloned repository root directory, or explicited like so:
```lua
package("mylib")
    add_urls("mylib.git")
    set_xmakefile(`"src/mylib/xmake.lua")
```

### Additional context

Here are some diagrams to help understand my problem
![Untitled Diagram-Page-2](https://github.com/xmake-io/xmake/assets/61986161/f7cdec91-b1b9-40e4-9890-1eddb44c320e)
![Untitled Diagram-Page-3](https://github.com/xmake-io/xmake/assets/61986161/99e59ea1-c638-4797-ac32-4ff82fd14da7)

**Comment by SirLynix** [comment]
I second this, it would greatly help for some packages.

It could be implemented using the xmake tools.
```lua
package("mylib")
    add_urls("mylib.git")
    
    on_load(function (package)
        import("package.tools.xmake").load(package, { requires = true }) -- gets all requires as private deps, except for public (used in add_deps("..", { public = true }))
    end)

**Comment by waruqi** [comment]
> An alternative is to declare dependencies in the package using , but then we would have to add a lot of branches to set the correct dependencies versions for each version of the library which would be worse for maintainability.

2.8.2 should have improved this, it will automatically sync the version of the deps to xmake.lua/add_requires.

2.8.2 should have fixed this, it will automatically sync the version of the deps to xmake.lua/add_requires

**Comment by waruqi** [comment]
> I second this, it would greatly help for some packages.
> 
> It could be implemented using the xmake tools.
> 
> ```lua
> package("mylib")
>     add_urls("mylib.git")
>     
>     on_load(function (package)
>         import("package.tools.xmake").load(package, { requires = true }) -- gets all requires as private deps, except for public (used in add_deps("..", { public = true }))
>     end)
> ```

If there is a conditional dependency of option + has_config + is_plat/is_arch + add_requires. We can't load them exactly.

```lua
if has_config("zlib") or is_plat("android") then
    add_requires("zlib")
end
```

**Comment by charlesseizilles** [comment]
> > An alternative is to declare dependencies in the package using , but then we would have to add a lot of branches to set the correct dependencies versions for each version of the library which would be worse for maintainability.
> 
> 2.8.2 should have improved this, it will automatically sync the version of the deps to xmake.lua/add_requires.
> 
> #3952

Well not it doesn't in my case. My lib's dependencies are considered private and I don't want the user to be able to change them. The issue is about versionning

**Comment by waruqi** [comment]
In addition, we are only extracting the package tar.gz in the on_install stage and then executing the tools.xmake process, if we load the information in the on_load stage, we need to download and extract all the package tar.gz in advance. But we haven't even loaded the package yet, so we can't extract the package tar.gz at this stage.

**Comment by waruqi** [comment]
> > > An alternative is to declare dependencies in the package using , but then we would have to add a lot of branches to set the correct dependencies versions for each version of the library which would be worse for maintainability.
> > 
> > 
> > 2.8.2 should have improved this, it will automatically sync the version of the deps to xmake.lua/add_requires.
> > #3952
> 
> Well no it doesn't in my case. My lib's dependencies are considered private and I don't want the user to be able to change them. The issue is about versionning

Even if a dependency is private and has not been modified by the user via add_requireconfs, xmake will also automatically synchronises the version information for all private dependencies

**Comment by charlesseizilles** [comment]
> > > > An alternative is to declare dependencies in the package using , but then we would have to add a lot of branches to set the correct dependencies versions for each version of the library which would be worse for maintainability.
> > > 
> > > 
> > > 2.8.2 should have improved this, it will automatically sync the version of the deps to xmake.lua/add_requires.
> > > #3952
> > 
> > 
> > Well no it doesn't in my case. My lib's dependencies are considered private and I don't want the user to be able to change them. The issue is about versionning
> 
> Even if a dependency is private and has not been modified by the user via add_requireconfs, xmake will also automatically synchronises the version information for all private dependencies

taking Lynix examples:
```lua
add_requires("fmt 8.0.1")

target("nzsl")
  add_packages("fmt")
  ...
```
```lua
package("nzsl")
  add_deps("fmt")

  on_install(function (package)
    import("package.tools.xmake").install(package)
  end)
```
```lua 
add_requires("nzsl")
```
does xmake know that it's fmt version 8.0.1 that needs to be linked in the user target, and not the latest?

**Comment by waruqi** [comment]
> > > > > An alternative is to declare dependencies in the package using , but then we would have to add a lot of branches to set the correct dependencies versions for each version of the library which would be worse for maintainability.
> > > > 
> > > > 
> > > > 2.8.2 should have improved this, it will automatically sync the version of the deps to xmake.lua/add_requires.
> > > > #3952
> > > 
> > > 
> > > Well no it doesn't in my case. My lib's dependencies are considered private and I don't want the user to be able to change them. The issue is about versionning
> > 
> > 
> > Even if a dependency is private and has not been modified by the user via add_requireconfs, xmake will also automatically synchronises the version information for all private dependencies
> 
> taking Lynix examples:
> 
> ```lua
> add_requires("fmt 8.0.1")
> 
> target("nzsl")
>   add_packages("fmt")
>   ...
> ```
> 
> ```lua
> package("nzsl")
>   add_deps("fmt")
> 
>   on_install(function (package)
>     import("package.tools.xmake").install(package)
>   end)
> ```
> 
> ```lua
> add_requires("nzsl")
> ```
> 
> does xmake know that it's fmt version 8.0.1 that needs to be linked in the user target, and not the latest?

xmake 2.8.2 know it, and it will use 8.0.1. you can try it.

**Comment by WubiCookie** [comment]
> I think you are trying to reverse sync from package xmake.lua/requires to user deps
I don't understand, what do you mean by that?

**Comment by waruqi** [comment]
package/foo

```lua
package("foo")
    add_deps("bar")         -------> B
    on_install(function (package)
        io.writefile("xmake.lua", [[
        add_requires("bar")         ---------> C
        target("foo")
            add_files("src/*.cpp")
            add_packages("bar")
]])
        import("package.tools.xmake").install(package)
    end)
```

user project:

```lua
add_requires("foo")      ----> A
target("test")
    add_files("src/*.cpp")
```

Current xmake will pass configs to package.

A (add_requireconfs) -> B (auto pass) -> C

User use `add_requireconfs` to pass configs/version A -> B
And xmake will pass configs/version B -> C automatically.

But I think that now the need for this issue is to reverse pass the dependency configuration and version.

C (auto pass) -> B -> A

**Comment by SirLynix** [comment]
Yes, the idea would be to greatly simplify making and maintaining packages for libraries using xmake.

However I don't really see how this would be possible without downloading the xmake.lua file first.

Also, maybe `xmake package` could be improved to provide the option to generate a single xmake.lua for all targets, instead of one per target.

**Comment by RimuruChan** [comment]
This is quite important when deps' version changed in different package version
